### DISCUSSION TITLE
tox4: plugin migration guidelines

### DISCUSSION THREAD

**Discussion by ssbarnea** [tox-dev/tox#1878] [General]
Lets try to use this thread for documenting how to port plugins to tox4.

One first question is around old `tox_configure` hook implementation which no longer exists.

**Comment by gaborbernat** [comment]
Can you detail exactly what's your use case/plugin functionality you want to migrate. Plugins interface is TBD at the moment, but the thought here is we now have a new paradigm so don't expect a 1 to 1 migration.

**Reply by ssbarnea** [reply]
I found this while writing one of the most basic plugins possible at https://github.com/ssbarnea/tox-bindep/blob/main/src/tox_bindep/hooks.py#L13

Basically I need to run some checks before the virtualenv is initialized, when I have access to tox own venv. This works fine with tox3 but I do not know how to make it work with tox4.

**Reply by gaborbernat** [reply]
Thanks, I'll have a think over this over the next week or so and get back to you when I've collected my thoughts.

**Comment by f3flight** [comment]
Hey folks, pretty much the same issue - I'm trying to migrate to tox 4 but I was using `tox_configure` in a custom internal plugin, which expected a fully populated config structure with env configs as well, and it would:

1. scan all envs and based on their basepython values, determine which python versions are required
2. replace all generic basepythonX.Y paths with custom paths, depending on version of python
3. find the highest python version used across version-specific envs (like py39 py310 etc)
4. replace basepython for all non-versioned envs with highest python version (i.e. use highest version for code quality and similar envs)

In `tox_configure`, I'd scan through config object and mutate some data inside. How do I do this now in tox 4?

**Reply by gaborbernat** [reply]
The base python configuration still controls the python version used. Furthermore, within the environment configuration hook you can both access and update this value. So that would be the path ahead.

**Reply by f3flight** [reply]
Hi Gabor! Thanks for some hints. I'd appreciate going into more detail:
- you say "the best python configuration", what do you mean by that?
- Within the environment configuration hook - do you refer to https://tox.wiki/en/latest/plugins.html#tox.plugin.spec.tox_add_env_config? I think this can cover `number 2` for version-specific environments like py39 py310 etc.
- How can I scan through all version-specific envs to find the highest requested version (`number 3`) before I hit any generic environments so that I could store that value and then use it to fulfill `number 4`  within `tox_add_env_config` when running into such environments?

**Reply by gaborbernat** [reply]
I don't follow what you're asking.

**Reply by f3flight** [reply]
tox allows to specify python version via environment name like so "py39", which will inclur python3.9. My plugin for tox v3 would look for all such environments in configuration, and find highest version of python. It would then use that version to replace basepython setting for all *other* environments, where it's not obvious which python version to use. This way users do not have to worry about basepython parameter at all in our custom environment, they only need to add pyXY into their environment list. The plugin takes care of setting basepython for all environments. How can I replicate this logic in tox v4?

**Reply by gaborbernat** [reply]
That sounds unsupported on purpose. Deviating from the built-in python discovery system of tox sounds like it would just introduce confusion for people expecting to follow the documented behavior. Perhaps you could do this with version 3, but I would consider that a significant bug of the system.

**Reply by f3flight** [reply]
Since this is an internal plugin, for internal users, and the workflow is documented, there is no confusion.  You mention python discovery mechanism, I looked at that (https://virtualenv.pypa.io/en/latest/extend.html#python-discovery) but I don't see how it can leverage tox configuration - config is not passed into it, so it cannot be "smart" about which python to provide to which environment, this mechanism can only provide the same python to all envs. I'm not even sure how can a custom discovery class replicate tox's standard behavior of providing pythonXY for "pyXY" env w/o explicit basepython setting (as described here - https://tox.wiki/en/latest/config.html#python-language-core-options `tox can set a default value from the environment name (e.g. py310 implies Python 3.10)`)

**Reply by gaborbernat** [reply]
If I'd join that company I'd be very confused. We don't plan to support your use case, you can fork the tool if you want, but I don't believe your approach is something reasonable or needs supporting by core. You're modifying the fundemantal concepts of the tool that I'm not comfortable with.

**Reply by f3flight** [reply]
Ok, then what kind of setup would you recommend instead?
How can I tell tox 4 where all my different python interpreters are so it could assign correct interpreters by itself? I have interpreters in different folders and these folders are not in PATH (by default, I can add of course).
How can I tell tox to use a specific python for all "other" environments, i.e. not py3X, while allowing it to assign automatically for py3X?

**Reply by gaborbernat** [reply]
for wherever the environment name does not specify a python version the python version tox is installed into will be used, so no need to tell it to tox, just install tox in the right environment 🤔

**Reply by gaborbernat** [reply]
See https://github.com/tox-dev/tox/blob/main/src/tox/tox_env/python/api.py#L138 and https://tox.wiki/en/latest/config.html#base_python

**Reply by f3flight** [reply]
I managed to achieve the desired result without using a plugin, by:
1. setting `base_python` manually in tox.ini of the project (not ideal but will do for now)
2. setting `ignore_base_python_conflict = true` in tox.ini to allow inference for `pyXY` envs even when `base_python` is set in `[testenv]`
3. adding all our python bin folders to `PATH` before starting tox, so it could infer version for `pyXY` environments

It would be nice if there was an option to prevent `base_python` setting from propagating to `pyXY` envs w/o using `ignore_base_python_conflict` because this flag also disables failure if pythonX.Y is not found on the host, which is not what I want - I want tox to fail in that case, instead of skipping.

**Reply by gaborbernat** [reply]
That combination might work but from my POV is unsupported. Why set base python and not just let the tox host environment decide the python version?

**Reply by f3flight** [reply]
We have multiple supported Python versions internally, all present at once on the host. We let the user choose which version they want/need to use for their project, i.e. a "primary" version; if we just choose latest or random all the time, we can cause unexpected failures when rolling out new Python on hosts, this will make builds unpredictable due to compatibility issues. Additionally, we distribute tox as a shiv zipapp, we dont want the user to install tox themselves, and we dont want the zipapp env used by tox to affect the user's project / dictate which Python version to use. We maintain zipapped tox, pdm and other tools of specific versions with a specific Python interpreter, and provide them as a curated-versioned "toolchain" for end users.

For libraries we also want to test on various versions, not just one, so we also want to leverage pyXY environments.

**Reply by gaborbernat** [reply]
So how do you determine what version is the default (aka non pyxy) 🤔

**Reply by f3flight** [reply]
Right now it is determined as highest from the list of pyXY envs in tox.ini. This avoids duplication, user needs to only declare pyXY (one or more) the app or library supports. This is with tox 3 with a plugin right now, but I need to migrate to v4.

**Reply by gaborbernat** [reply]
> Right now it is determined as highest from the list of pyXY envs in tox.ini.

Why? Why is this better than same version as tox running with?

**Reply by f3flight** [reply]
User's project may require an arbitrary version of Python to work. The code or dependencies may be incompatible with other Python versions. We cannot enforce on the user the version we use to provide tox as a zipapp. And we don't want our toolchain to depend on the version of Python a given user needs, which would require supporting toolchain bootstrapping for all python versions, or releasing multiple toolchain zipapps (tox37 tox38 pdm37 pdm38 etc).

**Reply by gaborbernat** [reply]
> User's project may require an arbitrary version of Python to work. The code or dependencies may be incompatible with other Python versions.

IMHO such envs should explicitely define the base python for these envs accordingly rather than depend on such arbitrary dynamic version discovery. Not specifying the base python is saying I don't care about the python version, any will do.

**Reply by f3flight** [reply]
I get your point, but at the same time I see specifying pyXY env(s) as explicit indication of supported python version, so I don't think the user needs to tell which version to use by default again separately via basepython setting.

**Reply by gaborbernat** [reply]
That would make the version of environments depend on other environments which in my POV is not a great place 🤔 Also applications don't tend to use `pyxy` format 😊 so this only works for libraries.

**Reply by gaborbernat** [reply]
> I see specifying pyXY env(s) as explicit indication of supported python version, so I don't think the user needs to tell which version to use by default again separately via basepython setting.

I guess this is where we disagree at core. I strongly believe otherwise, hence the tool is not designed to allow this behaviour (easily).

**Comment by ZenulAbidin** [comment]
I ported a plugin from using tox_configure in tox3 to tox_add_env_config in tox4. It checks for specific environment variables and some extra config options.

Tox doesn't seem to be complaining about that, so that's one avenue to explore.
